lib/models/package-info-cache/index.js

```diff
@@ -0,0 +1,751 @@
+'use strict';
+
+/*
+ * Performance cache for information about packages (projects/addons/"apps"/modules)
+ * under an initial root directory and resolving addon/dependency links to other packages.
+ */
+
+const fs = require('fs-extra');
+const path = require('path');
+const ErrorList = require('./error-list');
+const Errors = require('./errors');
+const PackageInfo = require('./package-info');
+const NodeModulesList = require('./node-modules-list');
+
+let realFilePathCache;
```

Comment:
You should have one package-info-cache each time you run 'cli'. That was an idea Stef and I discussed at the beginning, that you want to have a new cache each time you create a new CLI, so it works with the existing tests and test mechanism. The file and dir package cache needs to be cleared out each time you change CLI, so I just put them in module scope. They definitely do not want to be in global scope.

Reply:
Regarding the base comment, the problem with addon-discovery is that is was moving a lot of non-common code into a common location (e.g., the project-specific code) that really should have been in project.js to adhere to OO encapsulation principles. The generality was also slowing things down. I moved such things into project or addon where reasonable. discoverAddons() was actually in project.js and addon.js to start with, so I'd left them there for this work. Also, the set of things gathered in Addon is fairly different than that in Project, except for the handling of dependencies (which I take care of in PackageInfoCache) so I was trying to keep them separate rather than having a single method with various parameters located outside each - that's just more work for no particularly good reason, IMHO. For example, the test for excluding 'ember-cli' in dependencies is Addon-specific, not in Project. The test for including CLI is project only, not addon. Given the differences, I'd really prefer to leave things as they are for this part.